source/s3_buffer_pool.c

```diff
@@ -0,0 +1,228 @@
+    uint8_t *offset_ptr;
+
+    size_t alloc_count;
+};
+
+struct aws_s3_buffer_pool *aws_s3_buffer_pool_new(
+    struct aws_allocator *allocator,
+    size_t block_size,
+    size_t max_mem_usage) {
+    struct aws_s3_buffer_pool *buffer_pool = aws_mem_calloc(allocator, 1, sizeof(struct aws_s3_buffer_pool));
+
+    AWS_FATAL_ASSERT(buffer_pool != NULL);
+
+    buffer_pool->base_allocator = allocator;
+    buffer_pool->block_size = block_size;
+    buffer_pool->current_mem_usage = 0;
+    buffer_pool->max_mem_usage = max_mem_usage;
+    if (aws_mutex_init(&buffer_pool->mutex)) {
+        goto on_error;
+    }
+
+    aws_array_list_init_dynamic(&buffer_pool->blocks, allocator, s_min_num_blocks, sizeof(struct s3_buffer_pool_block));
+
+    return buffer_pool;
+
+on_error:
+    aws_mem_release(allocator, buffer_pool);
+    return NULL;
+}
+
+void aws_s3_buffer_pool_destroy(struct aws_s3_buffer_pool *buffer_pool) {
+    if (buffer_pool == NULL) {
+        return;
+    }
+
+    for (size_t i = 0; i < aws_array_list_length(&buffer_pool->blocks); ++i) {
+        struct s3_buffer_pool_block *block;
+        aws_array_list_get_at_ptr(&buffer_pool->blocks, (void **)&block, i);
+
+        AWS_FATAL_ASSERT(block->alloc_count == 0 && "Allocator still has outstanding blocks");
+        aws_mem_release(buffer_pool->base_allocator, block->block_ptr);
+    }
+
+    aws_array_list_clean_up(&buffer_pool->blocks);
+
+    aws_mutex_clean_up(&buffer_pool->mutex);
+    struct aws_allocator *base = buffer_pool->base_allocator;
+    aws_mem_release(base, buffer_pool);
+}
+
+void aws_s3_buffer_pool_trim(struct aws_s3_buffer_pool *buffer_pool) {
+    size_t swap_count = 0;
+
+    for (size_t i = 0; i < aws_array_list_length(&buffer_pool->blocks); ++i) {
+        struct s3_buffer_pool_block *block;
+        aws_array_list_get_at_ptr(&buffer_pool->blocks, (void **)&block, i);
+
+        if (block->alloc_count == 0) {
+            size_t swap_loc = aws_array_list_length(&buffer_pool->blocks) - 1 - swap_count;
+            if (swap_loc == i) {
+                break;
+            } else {
+                aws_array_list_swap(&buffer_pool->blocks, i, swap_loc);
+            }
+        }
+    }
+
+    for (size_t i = 0; i < swap_count; ++i) {
+        buffer_pool->current_mem_usage -= buffer_pool->block_size;
+        aws_array_list_pop_back(&buffer_pool->blocks);
+    }
+}
+
+static struct aws_s3_pooled_buffer s_primary_acquire(struct aws_s3_buffer_pool *buffer_pool, size_t size) {
+    uint8_t *alloc_ptr = NULL;
+    aws_mutex_lock(&buffer_pool->mutex);
+
+    for (size_t i = 0; i < aws_array_list_length(&buffer_pool->blocks); ++i) {
+        struct s3_buffer_pool_block *block;
+        aws_array_list_get_at_ptr(&buffer_pool->blocks, (void **)&block, i);
+
+        if (block->alloc_count == 0) {
+            block->offset_ptr = block->block_ptr + size;
+            alloc_ptr = block->block_ptr;
+            block->alloc_count += 1;
+            break;
+        }
```

Comment:
trivial/debatable: you could either eliminate this if-block if you reset the offset_ptr from the release() function.

or you could just have this if-block reset the offset_ptr, and let the next if-block handle everything else